### PR TITLE
Agent Argument Parsing

### LINKED ISSUE #2034: Allow command-line args for agent configuration
Many folks who have been trying to run the agents find it unintuitive to specify ENV vars. It seems more natural to folks to configure via command-line args. Many node agents like geth do it that way.

While the config library we use does not directly support command-line arguments, the `Source` trait seems flexible enough to allow a thin wrapper over an existing CMD-line argument library to be provided a map of key-value-pairs to loading to the config library. This would seem like an easy enough way to get command line argument support.

Relevant Discord threads:
- https://discord.com/channels/935678348330434570/1077336620941987951/1077529971242840117

### PR DESCRIPTION
**Pull request by mattiecnvr**
### Description

All configs can now be set by arguments as well as env and config files. The arguments will take the highest priority and overwrite the others.

We should also update the docs to include this new support when it gets released, but since it is an entirely new feature it also can be a stealth release which we document after the fact.

### Related issues

- Fixes #2034 

### Backward compatibility

_Are these changes backward compatible?_

Yes

_Are there any infrastructure implications, e.g. changes that would prohibit deploying older commits using this infra tooling?_

None

### Testing

_What kind of testing have these changes undergone?_

Manual
Unit Tests

### CODE REVIEW

File: rust/hyperlane-base/src/settings/loader/mod.rs [hunk 1]
```diff
@@ -72,10 +75,11 @@ where
                 .separator("_")
                 .source(Some(filtered_env)),
         )
+        .add_source(CommandLineArguments::default().separator("."))
```
Comment (mattiecnvr) [inline rust/hyperlane-base/src/settings/loader/mod.rs:78]: LMK if you think we should use something else such as `-`.

This currently looks like `--chains.ethereum.conenction.url="https://example.com"` but we could also do something like `--chains-ethereum-conenction-url="https://example.com"`.

File: rust/hyperlane-core/src/config/config_path.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    /// Get the expected command line argument name
+    pub fn arg_name(&self) -> String {
+        let name = self
+            .0
+            .iter()
+            .map(|s| s.as_str().to_case(Case::Camel))
+            .join(".");
+        format!("--{name}")
```
Comment (mattiecnvr) [inline rust/hyperlane-core/src/config/config_path.rs:79]: Something to note here is that we might encourage people to use camel cased names for readability but ultimately they won't be displayed here.

File: rust/utils/run-locally/src/main.rs [hunk 3]
```diff
@@ -142,23 +142,29 @@ fn main() -> ExitCode {
     };
 
     let relayer_env = hashmap! {
```
Comment (mattiecnvr) [inline rust/utils/run-locally/src/main.rs:144]: No net changes to the config, just trying to make sure the args correctly are parsed and override the env.

File: rust/hyperlane-core/src/config/config_path.rs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    fn add(self, rhs: ConfigPath) -> Self::Output {
+        self.merge(&rhs)
+    }
+}
+
+impl ConfigPath {
+    /// Add a new part to the path.
+    pub fn join(&self, part: impl Into<String>) -> Self {
+        let part = part.into();
+        debug_assert!(!part.contains('.'));
```
Comment (tkporter) [inline rust/hyperlane-core/src/config/config_path.rs:38]: wondering what situation this is supposed to catch and if it should be something we want in the optimized build?

(i know this is just a moved bit of code that already existed)

**Review by tkporter — APPROVED** [review]
awesome! lgtm

File: rust/hyperlane-core/src/config/config_path.rs (hunk 4 above)
Comment (mattiecnvr) [reply rust/hyperlane-core/src/config/config_path.rs:38]: It catches a case that isn't going to break anything, but that would be nice to catch during debug. Basically we should not have the path include "." but have different segments instead.

**Comment by nambrot** [comment]
@mattiecnvr Do you want me to add another ticket for the docs update or you want to do that in the context of #2113 maybe?

**Pull request merged**